src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -494,6 +494,12 @@ InterpreterSelectQuery::InterpreterSelectQuery(
         query_info.additional_filter_ast = parseAdditionalFilterConditionForTable(
             settings.additional_table_filters, joined_tables.tablesWithColumns().front().table, *context);
 
+
+    if (!query.final() && storage && storage->forceQueryWithFinal())
+    {
+        query.setFinal();
```

Comment:
There is already a similar method, named `needRewriteQueryWithFinal`.

Reply:
Reading the docs and the implementation I see that methods is for a different use-case. Are you suggesting I incorporate the logic of this method onto the other one and update the docs?

Reply:
I have tried to incorporate the logic of the `forceQueryWithFinal` into the `needRewriteQueryWithFinal`, but it didn't work as expected. The simplest test case (table with \setting turned on and select query) fails. It is because `needRewriteQueryWithFinal` is being called  after the `TreeRewriter::analyzeSelect` call, whereas `forceQueryWithFinal` was being called before it. `TreeRewriter::analyzeSelect` calls `query.final` to decide if `optimize_trivial_count` should be on or off. In a nutshell, at that point, select final is not yet set and then `optimize_trivial_count` takes place. I can not move the `needRewriteQueryWithFinal` call up because it seems to depend on `TreeRewriter::analyzeSelect` to get the requiredSourceColumns